docs/DifferentiableProgrammingImplementation.md

```diff
@@ -0,0 +1,1012 @@
+
+In the sections below, let us dig into derivative code generation for the following function-to-differentiate (“original function”):
+
+```swift
+@differentiable
+func foo(_ x: Float) -> Float {
+    return sin(x) * cos(x)
+}
+
+// Simplified SIL pseudocode.
+sil @foo : $(Float) -> Float {
+bb0(%x):
+  %y1 = apply @sin(%x)
+  %y2 = apply @cos(%x)
+  %y3 = apply @mul(%y1, %y2)
+  return %y3
+}
+```
+
+## JVP and differential generation
+
+### Overview
+
+Generated JVP functions are clones of the original function, with function applications replaced with JVP applications. Each JVP application returns the original result of the application in addition to a differential value (named “callee differential” to distinguish it from the “generated differential function”).
+
+The generated differential function takes the partial derivative with respect to inputs and returns the partial derivative with respect to the output. There is a mapping between active values in the original function and their tangent values (i.e. partial derivatives) in the differential function. The differential captures the “callee differentials” from the JVP, and replaces original function applications using original values with “callee differential” applications using tangent values.
+
+Finally, the JVP returns a tuple of the original result and the generated differential function.
+
+```swift
+// JVP: replaces all function applications with JVP applications.
+sil @jvp_foo : $(Float) -> (Float, (Float) -> Float) {
+bb0(%x):
+  (%y1, %df_sin) = apply @jvp_sin(%x)
+  (%y2, %df_cos) = apply @jvp_cos(%x)
+  (%y3, %df_mul) = apply @jvp_mul(%y1, %y2)
+  // Return tuple of original result and differential.
+  return (%y3, { %dx in
+    %dy1 = apply %df_sin(%dx)
+    %dy2 = apply %df_cos(%dx)
+    %dy3 = apply %df_mul(%dy1, %dy2)
+    return %dy3
+  })
+}
+```
+
+### Visiting the original function to create a JVP
+
+In the Swift compiler, we have a class called [`JVPEmitter`](https://github.com/apple/swift/blob/tensorflow/lib/SILOptimizer/Mandatory/Differentiation.cpp) which subclasses [`TypeSubstCloner`](https://github.com/apple/swift/blob/master/include/swift/SIL/TypeSubstCloner.h). This uses the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) to visit the original function and generate the JVP function. Taking a look at the `JVPEmitter` class, there are methods like `visitApplyInst`, `visitStructExtractInst`, etc. Each of these methods visit an instruction that is important in the generation of the JVP and differential, and emits a newly mapped version. We handle each type of instruction differently, explained below.
```

Comment:
This is out of date with the current state of main.  I think that we should not be referencing anything on the tensorflow branch.

Reply:
Could you replace the link with the correct one then?
```suggestion
In the Swift compiler, we have a class called [`JVPEmitter`](https://github.com/apple/swift/blob/main/lib/SILOptimizer/Mandatory/Differentiation.cpp) which subclasses [`TypeSubstCloner`](https://github.com/apple/swift/blob/master/include/swift/SIL/TypeSubstCloner.h). This uses the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) to visit the original function and generate the JVP function. Taking a look at the `JVPEmitter` class, there are methods like `visitApplyInst`, `visitStructExtractInst`, etc. Each of these methods visit an instruction that is important in the generation of the JVP and differential, and emits a newly mapped version. We handle each type of instruction differently, explained below.
```